Clarify/cleanup backend upgrade (#3214)
chef/chef-server: docs-chef-io/content/server/upgrade_server_ha_v2.md

#### Identify the Index Elasticsearch Version

Find the version of Elasticsearch that created the indexes. Run the following command on a back-end node and find value of the `created_string:` in the output:

```bash
curl -XGET localhost:9200/chef/_settings/index.version.created*?pretty\&human
```

> #### Check for index health

Reviewer: ```suggestion
#### Check for Index Health
```

> All indexes should be green from a frontend server. There should be no other indexes present than the one named "chef", and definitely no non-green indexes. The command should have a return code of 1. Investigate by removing the grep otherwise.

Reviewer: ```suggestion
Only "chef" indexes should be present, and all indexes should have a green health status. The following command should have a return code of 1:
```

```bash
curl -s -S "localhost:9200/_cat/indices?h=health,status,index" | grep -v '^green'
```

> 

Reviewer: ````suggestion

Investigate the health status of your indexes by running the same command without grep:

```bash
curl -s -S "localhost:9200/_cat/indices?h=health,status,index"
```

````

    chef-backend-ctl set-cluster-failover off
    ```

> 1. Proceed to rebuilding the "chef" index [Rebuild Index]({{< relref "#rebuild-the-chef-index" >}})

Reviewer: ```suggestion
1. [Rebuild the "chef" Index]({{< relref "#rebuild-the-chef-index" >}})
```

For indexes created by Elasticsearch version 5.6.16 (`"created_string" : "5.6.16"`) or later, follow these steps:

1. Identify the node with the **leader** role using the `chef-backend-ctl cluster-status` command:

    ```none
    Name       IP              GUID                              Role      PG        ES
    backend-1  192.168.33.215  dc0c6ea77a751f94037cd950e8451fa3  leader    leader    not_master
    backend-2  192.168.33.216  008782c59d3628b6bb7f43556ac0c66c  follower  follower  not_master
    backend-3  192.168.33.217  1af654172b1830927a571d9a5ba7965b  follower  follower  master
    ```

    In this example, `backend-1` is the **leader** node, as indicated by its role in the **Role** column.

1. Install the new Chef Backend package on all nodes in the cluster:

    RHEL and CentOS:

    ```bash
    yum install PATH_TO_FILE.rpm
    ```

    Debian and Ubuntu:

    ```bash
    dpkg -i PATH_TO_FILE.deb
    ```

1. Upgrade on follower nodes first:

    ```bash
    chef-backend-ctl upgrade
    ```

1. Upgrade the leader node with the `--failover` option:

    ```bash
    chef-backend-ctl upgrade --failover
    ```
    

>     If you see lines with the following output in /var/log/chef-backend/elasticsearch/current just after the attempted upgrade or a problem in the output from the upgrade command
>     
>     ```bash
>     unknown setting [path.scripts]
>     ```
>     
>     Then the system has not upgraded fully. Make sure it is not still the leader, then if it is not but the ES cluster is green

Reviewer: ```suggestion
    The system has not upgraded fully if you see `unknown setting [path.scripts]` in `/var/log/chef-backend/elasticsearch/current` just after the attempted upgrade or errors in the upgrade command output.
    Make sure the node is not the leader and the Elasticsearch cluster health status is green, then upgrade the node again:
```

    
    ```bash
    chef-backend-ctl upgrade
    ```
    
    Afterwards, the original leader will have upgraded completely.

> 1. To continue the upgrade, let's move to the Chef Infra Server front-end nodes using this back-end cluster, see [Upgrade Front-ends Associated with a Chef Backend Cluster]({{< relref "install_server_ha.md/#upgrading-chef-infra-server-on-the-frontend-machines" >}}).

Reviewer: ```suggestion
1. Next, [upgrade the Chef Infra Server frontend nodes using this backend cluster]({{< relref "#upgrading-chef-infra-server-on-the-frontend-machines" >}}).
```

> 1. We strongly recommend rebuilding the "chef" index version to after the upgrade is complete on both backends and frontends by following these steps [Rebuild Index]({{< relref "#rebuild-the-chef-index" >}})

Reviewer: ```suggestion
1. We strongly recommend [rebuilding the "chef" index]({{< relref "#rebuild-the-chef-index" >}}) after the upgrade is complete on both backend and frontend nodes.
```

> ## Rebuild the chef index

Reviewer: ```suggestion
## Rebuild the Chef Index
```

> 1. Set cluster failover to off on a backend

Reviewer: ```suggestion
1. Prevent the cluster from failing-over while the index is being rebuilt:
```

   ```bash
   chef-backend-ctl set-cluster-failover off
   ```

1. Stop the Elasticsearch services on all backend nodes:

   ```bash
   sudo chef-backend-ctl stop elasticsearch
   ```

> 1. Delete the current Elasticsearch index. This step ensures that Elasticsearch creates a new index with the upgraded Elasticsearch version. Use the command:

Reviewer: ```suggestion
1. Delete the current Elasticsearch index. Elasticsearch will create a new index with the upgraded Elasticsearch index version. Use the command:
```

For indexes created by Elasticsearch version 5.6.16 (`"created_string" : "5.6.16"`) or later, follow these steps:

1. Identify the node with the **leader** role using the `chef-backend-ctl cluster-status` command:

    ```none
    Name       IP              GUID                              Role      PG        ES
    backend-1  192.168.33.215  dc0c6ea77a751f94037cd950e8451fa3  leader    leader    not_master
    backend-2  192.168.33.216  008782c59d3628b6bb7f43556ac0c66c  follower  follower  not_master
    backend-3  192.168.33.217  1af654172b1830927a571d9a5ba7965b  follower  follower  master
    ```

    In this example, `backend-1` is the **leader** node, as indicated by its role in the **Role** column.

1. Install the new Chef Backend package on all nodes in the cluster:

    RHEL and CentOS:

    ```bash
    yum install PATH_TO_FILE.rpm
    ```

    Debian and Ubuntu:

    ```bash
    dpkg -i PATH_TO_FILE.deb
    ```

1. Upgrade on follower nodes first:

    ```bash
    chef-backend-ctl upgrade
    ```

1. Upgrade the leader node with the `--failover` option:

    ```bash
    chef-backend-ctl upgrade --failover
    ```
    

>     If you see lines with the following output in /var/log/chef-backend/elasticsearch/current just after the attempted upgrade or a problem in the output from the upgrade command
>     
>     ```bash
>     unknown setting [path.scripts]
>     ```
>     
>     Then the system has not upgraded fully. Make sure it is not still the leader, then if it is not but the ES cluster is green
>     
>     ```bash
>     chef-backend-ctl upgrade
>     ```
>     
>     Afterwards, the original leader will have upgraded completely.

Reviewer: ```suggestion
    {{< note spaces = 4 >}}

    If you see `unknown setting [path.scripts]` in `/var/log/chef-backend/elasticsearch/current` after the attempted upgrade, or a problem in the output from the upgrade command, then the system has not upgraded fully. Make sure the node is not the leader and the Elasticsearch cluster health status is green, then upgrade the node again:

    ```bash
    chef-backend-ctl upgrade
    ```

    Afterwards, the original leader will have upgraded completely.

    {{< /note >}}
```
